osutil/resolve_path.go

```diff
@@ -0,0 +1,136 @@
+ *
+ */
+
+package osutil
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+func dumbJoin(a, b string) string {
+	if strings.HasSuffix(a, "/") {
+		return a + b
+	} else {
+		return a + "/" + b
+	}
+}
+
+func resolvePathInSysrootRec(sysroot, path string, validInAnySysroot bool, symlinkRecursion int) (string, error) {
+	if path == "" || path == "/" {
+		// Relative paths are taken from sysroot
+		return "/", nil
+	}
+
+	if strings.HasSuffix(path, "/") {
+		path = path[:len(path)-1]
+	}
+
+	dir, file := filepath.Split(path)
+	resolvedDir, err := resolvePathInSysrootRec(sysroot, dir, validInAnySysroot, symlinkRecursion)
+	if err != nil {
+		return "", err
+	}
+	if file == "" {
+		return resolvedDir, nil
+	}
+	if file == "." {
+		return resolvedDir, nil
+	}
+	if file == ".." {
+		if validInAnySysroot && (resolvedDir == "/") {
+			return "", fmt.Errorf("invalid escaping path")
+		}
+		upperDir, _ := filepath.Split(resolvedDir)
+		return upperDir, nil
+	}
+
+	fileInResolvedDir := dumbJoin(resolvedDir, file)
+
+	realPath := dumbJoin(sysroot, fileInResolvedDir)
+	st, err := os.Lstat(realPath)
+	if err != nil {
+		return "", err
+	}
+
+	if st.Mode()&os.ModeSymlink != 0 {
+		if symlinkRecursion < 0 {
+			return "", fmt.Errorf("maximum recursion reached when reading symlinks")
+		}
+		target, err := os.Readlink(realPath)
+		if err != nil {
+			return "", err
+		}
+		if filepath.IsAbs(target) {
+			if validInAnySysroot {
+				return "", fmt.Errorf("invalid absolute symlink")
+			}
+			return resolvePathInSysrootRec(sysroot, target, validInAnySysroot, symlinkRecursion-1)
+		} else {
+			return resolvePathInSysrootRec(sysroot, dumbJoin(resolvedDir, target), validInAnySysroot, symlinkRecursion-1)
+		}
+	}
+
+	return fileInResolvedDir, nil
+}
+
+// ResolvePathInSysroot resolves a path within a sysroot
+//
+// In a sysroot, abolute symlinks should be relative to the sysroot
+// rather than `/`. Also paths with multiple `..` that would escape
+// the sysroot should not do so.
+//
+// The path must point to a file that exists.
+//
+// Example 1:
+//   - /sysroot/path1/a is a symlink pointing to /path2/b
+//   - /sysroot/path2/b is a symlink pointing to /path3/c
+//   - /sysroot/path3/c is a file
+//     ResolvePathInSysroot("/sysroot", "/path1/a") will return "/path3/c"
+//
+// Example 2:
+//   - /sysroot/path1/a  is a symlink pointing to ../../../path2/b
+//   - /sysroot/path2/b  is a symlink pointing to ../../../path3/c
+//   - /sysroot/path3/c  is a file
+//     ResolvePathInSysroot("/sysroot", "../../../path1/a") will return "/path3/c"
+//
+// Example 3:
+//   - /sysroot/path1/a is a symlink pointing to /path2/b
+//   - /sysroot/path2/b does not exist
+//     ResolvePathInSysroot("/sysroot", "/path1/a") will fail (IsNotExist)
+//
+// The return path is the path within the sysroot. filepath.Join() has
+// to be used to get the path in the sysroot.
+func ResolvePathInSysroot(sysroot, path string) (string, error) {
```

Comment:
This function not actually used currently, is that correct? We only need the `NoEscape` version?

Reply:
We only need the NoEscape for this case. But because it is a useful function and the algorithm is the same I wanted to make the general case to work.